### PR TITLE
Add concurrency support for more than one thread

### PR DESCRIPTION
**Pull request by vicary**
## What did you implement:

Adding concurrency support for more than one thread.

Conditionally supersedes #680 if `release/5.4.0` is more preferable than `master`.

## How did you implement it:

Please see https://github.com/serverless-heaven/serverless-webpack/pull/517#issuecomment-653547396, basically I did two changes:

1. Replacing `serizliedCompile: Boolean` with `concurrency: Number`
2. Translates `serializedCompile: true` to `concurrency: 1` for backward compatibility

## How can we verify it:

Use the config below when packaging:

```
package:
  individually: true

custom:
  webpack:
    concurrency: 5
```

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** Yes
***Is it a breaking change?:*** No

### CODE REVIEW

**Comment by coyoteecd** [comment]
@miguel-a-calles-mba @j0k3r can you have a look and merge this? I am using vicary's fork for half a year and would love to switch back to the official package.

**Comment by j0k3r** [comment]
@coyoteecd we are in the process of cleaning up stuff to release 5.4.0 and this will be part of it

**Comment by j0k3r** [comment]
Sorry I made a mistake while merging `release/5.4.0` into master and removed it. The removal automatically closed that PR. It should now be rebased against the master and fix conflicts.

**Comment by vicary** [comment]
Not sure if I'm doing it wrong, but I found myself learning the backstory every conflict or two in the process of rebasing. This takes a day or two, could only do this on weekends.

**Comment by vicary** [comment]
@j0k3r During my rebase I found that the current master has these patterns.

```js
// tests/packageModules.test.js
        return BbPromise.each([ '1.18.1', '2.17.0', '10.15.3' ], version => {
          getVersionStub.returns(version);
          return expect(module.packageModules()).to.be.fulfilled.then(() => BbPromise.all([]));
        }).then(() =>
          BbPromise.each([ '1.17.0', '1.16.0-alpha', '1.15.3' ], version => {
            getVersionStub.returns(version);
            return expect(module.packageModules()).to.be.fulfilled.then(() => BbPromise.all([]));
          })
        );
```

It has a strong code-smell here `.then(() => BbPromise.all([]));` which should either not there, or some tests are missed during rebase.

Not meant to make things more complicated than it should, please feel free to ignore this if it's intended.

I'll be rebasing to master as-is, just a friendly heads-up.

**Comment by j0k3r** [comment]
@vicary I've no idea but it seems useless 🤔 Have you tried to remove the useless part `.then(() => BbPromise.all([]));` to see if tests are still green?

**Comment by vicary** [comment]
> @vicary I've no idea but it seems useless 🤔 Have you tried to remove the useless part `.then(() => BbPromise.all([]));` to see if tests are still green?

They must produce the same result, because it does not add any logics besides an empty promise resolution.

From the rebase conflicts I can see there was other assertions inside the arrays. I am not sure if these sub-assertions are meant to be removed, if so, the code segment can be safely removed.

**Comment by vicary** [comment]
The rebase is finished, this should be ready for merging.

File: lib/Configuration.js [hunk 1]
```diff
@@ -15,7 +15,7 @@ const DefaultConfig = {
   packagerOptions: {},
   keepOutputDirectory: false,
   config: null,
-  serializedCompile: false
+  concurrency: Infinity
```
Comment (j0k3r) [inline lib/Configuration.js:18]: So the default value is infinity? Shouldn't this be _to the number of available cores_?

**Comment by j0k3r** [comment]
Awesome job for the rebase @vicary 👏

File: lib/Configuration.js (hunk 1 above)
Comment (vicary) [reply lib/Configuration.js]: You're correct.

File: lib/validate.js [hunk 2]
```diff
@@ -190,10 +190,11 @@ module.exports = {
       // In case of individual packaging we have to create a separate config for each function
       if (_.has(this.serverless, 'service.package') && this.serverless.service.package.individually) {
         this.multiCompile = true;
-        this.serializedCompile = this.configuration.serializedCompile;
         this.options.verbose &&
           this.serverless.cli.log(
-            `Using ${this.serializedCompile ? 'serialized' : 'multi'}-compile (individual packaging)`
+            `Using ${
+              this.configuration.concurrency !== Infinity ? 'concurrent' : 'multi'
```
Comment (j0k3r) [inline lib/validate.js:196]: Shouldn't this be updated as well? Since concurrency default value is no more `Infinity`

**Review by j0k3r — COMMENTED** [review]
Also, I think some tests should be updated as well about `{ concurrency: Infinity }`. Don't you think?

File: lib/validate.js (hunk 2 above)
Comment (vicary) [reply lib/validate.js:196]: Before `serializedCompile` is available, `serverless-webpack` blasts everything at once in the same process, that's where the `Infinity` comes from.

**Comment by vicary** [comment]
> Also, I think some tests should be updated as well about `{ concurrency: Infinity }`. Don't you think?

I agree that using `Infinity` may causes confusions. In fact `compile.test.js` is largely testing if the compiler has even run once, `concurrency` config there is mostly for cosmetic reasons.

To reduce ambiguity, how about we change the tests to use `1` instead?

File: lib/validate.js (hunk 2 above)
Comment (vicary) [reply lib/validate.js:196]: I just had a deeper look on this. When `package.individually === true` it set `multiCompile = true`, my searches tells me this property is no longer used anywhere else. In fact the original `Infinity` behavior must be explicitly forced via a `${file:...}` reference in `serverless.yml`.

Do you think it's a good idea to remove all instance of *multiCompile* and change the log to the following format instead? 
```javascript
`Individually packaging with concurrency at ${this.configuration.concurrency} entries a time.`
```

Comment (j0k3r) [reply lib/validate.js:196]: Looks like `multiCompile` was removed here https://github.com/serverless-heaven/serverless-webpack/pull/517
So yes, good idea to change the log 👍

Comment (vicary) [reply lib/validate.js]: All done

**Review by j0k3r — APPROVED** [review]
Good job!

**Pull request merged**